### DISCUSSION TITLE
esp32cam change url stream

### DISCUSSION THREAD

**Discussion by defaliz** [arendst/Tasmota#14588] [Q&A] [answered]
Hi,
i use tasmota for several devices and i like it wery well ! thank arendst !
i use esp32cam over internet with nginx proxy.. so url for looking stream is not with local ip but with internet ip
i'd like to change stream ip by compiling tasmota cam firmware...
change <img src='http://private_ip:81/stream'> with my public ip..
is it possible by modifying user_config_override.h ?
thank for your help
regards
Alain

**Comment by barbudor** [comment, marked as answer]
Hi @defaliz 
Unfortunately not today.
A #define could be introduced but that would not be a very nice solution. It would even better if we could have a relative link instead of an absolute one.
As of today the HTML code introduced is:
```
<p></p><center><img src='http://%_I:81/stream' alt='Webcam stream' style='width:99%%;'></center><p></p>
```
Where `%_I` is replace by the Wifi IP address
Unfortunately I've not yet found a relative syntax that could allow to switch to another port
Like `:81/stream` or `http://:81/stream` doesn't work.
If you have an idea...

**Reply by defaliz** [reply]
thank Mark
maybe if i try scripting (compiling .bin before to enable script)
here is an example from tasmota website

the webcam stream can be specified by the following line
lip is a system variable containing the local device ip
`"&<br>"
"&<img src="http://%lip%:81/stream" style="width:%w%px;height:%h%px">"
"&<br><center>webcam stream" `

if i change %lip% by my public address, maybe that could work..

**Reply by sfromis** [reply]
Well, if you have a public IP address, which is mapped to your internal IP address, I do not think that Tasmota needs to be involved for this, as I'd expect it to work if you directly access `http://pub.lic.ip.address:81/stream` 
The only thing you lose is accessing the Tasmota web server html page including the link generated by Tasmota.

Of course, beware of the bad security implications of opening **any** IP address to the big bad internet.

**Reply by barbudor** [reply]
@sfromis tasmota main web page explicitly have a link to the stream URL that is absolute using the wifi address.
As said above, because of the port change there is no way to make it relative.

**Reply by sfromis** [reply]
Yes, that's specifically why I suggested to use the stream link *directly* and warned that the html front page would not be usable, @barbudor

**Reply by defaliz** [reply]
ok, for sure i allready use the stream link on port 81, no problem it's ok, BUT,
i prefer to have the streaming in the man web page.. and that don't work because the ip stream is the local ip..
no problem about security from internet, i use my own proxy and login/password protect..
i think the only thing is to change stream url, by compiling or scripting..

**Reply by sfromis** [reply]
You can just create a small web page to host the stream link. If you need the standard web UI too, you still have port 80, just not having the stream accessible directly on the Tasmota front page.

**Reply by defaliz** [reply]
yes sfromis, it's a good idea... iframe for web ui page and iframe for url stream ?

**Reply by sfromis** [reply]
Here's a hack for you..... Configure a relay, and issue the command:
`WebButton1 <img src="http://pub.lic.ip.address:81/stream">`
This makes the button include the stream as an image.

Using a couple of frames also makes sense, of course :-)

**Reply by defaliz** [reply]
great ! very good idea the webbutton ! i tried asap and tell you..
thank a lot !

**Reply by defaliz** [reply]
hi sfromis, you save my life ! works perfectly !
![image](https://user-images.githubusercontent.com/41428343/150945208-aec9b9b4-e3a9-4974-9903-ccf4b7782c54.png)
now i can have stream in the main webpage over the internet !
thanks and have a good day
Alain

**Reply by sfromis** [reply]
Not very pretty, but works :-) 

If you want, you can tone down the blue by setting WebColor. And maybe `SetSensor127 0` to get rid of that useless `ESP32 Temperature`.

**Reply by defaliz** [reply]
yes, again a good idea..
many thanks

**Comment by Nemobi** [comment]
One way would be to let the browser resolve the hostname using `window.location.hostname`. [This StackOverflow question](https://stackoverflow.com/questions/6016120/relative-url-to-a-different-port-number-in-a-hyperlink) suggests a couple of solutions, and I think the most succinct would be something like this:
```
<script>document.write(\"<p></p><center><img src='http://\"+window.location.hostname+\":81/stream' alt='Webcam stream' style='width:99%%;'></center><p></p>\")</script>
```
